backend/instancemgmt/instance_manager.go

```diff
@@ -119,7 +121,11 @@ func (im *instanceManager) Get(ctx context.Context, pluginContext backend.Plugin
 		}
 
 		if disposer, valid := ci.instance.(InstanceDisposer); valid {
-			disposer.Dispose()
+			time.AfterFunc(disposeTTL, func() {
+				disposer.Dispose()
+				activeInstances.Dec()
+			})
+		} else {
 			activeInstances.Dec()
 		}
```

Comment:
Yes, this should resolve a bug where earlier we didn't decrease the counter if `Dispose` was not implemented by a plugin. So looking at the metric right now you might get the impression that active instances is steadily increasing, but in reality they might not.

> I was also thinking maybe we can keep a counter of how many times an instance is retrieved, and decrement it when it's being "released". So adding a Release method to the instance manager alongside Get. Then the caller is responsible for defer-calling Release. This would allow us to call Dispose when the counter reaches 0 (so nobody is using a given instance). However this complicates things, changes the interface and breaks plugins that use manual instance management (but we are discouraging that), so probably not worth it, but wanted to hear your thoughts on this as well

Yes, me and @wbrowne discussed/considered something like this but it complicates things quite a lot and might require even more looking to be implemented and potentially decreasing the performance. If these changes are shown to not work as expected we might need to consider something like you suggest. Does that make sense?

Reply:
I agree. I also feel like this should be enough for now and doesn't change too many things. Let's go for it!